csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1385,62 +1409,377 @@ class AllocationReuseModifier : private kir::ExprMutator {
+//! near the end of the Fusion, then with this approach we will not be able to
+//! re-use its memory in the middle, which might be wasteful.
+class StackBasedSharedMemAllocator : kir::IrVisitor {
  public:
+  StackBasedSharedMemAllocator(const AllocationInfoMap& allocation_info_map)
+      : allocation_info_map_(allocation_info_map) {}
+
   void allocate(std::vector<Expr*>& exprs) {
+    recordEvents();
+
+    // Traverse expressions: reclaim memory when we pass a blockSync, append to
+    // waiting_to_push_ when we pass an Allocate
     handle(exprs);
+
+    // This is done whenever we pass a syncing op, but we need to do it again in
+    // case there are some allocations waiting around to be allocated.
+    sortPushAndAssignWaiting();
   }
 
  private:
   using kir::IrVisitor::handle;
 
-  static Val* alignExpr(Val* addr, int64_t alignment = 16) {
-    if (alignment == 1) {
-      return addr;
+  void dispatch(Expr* expr) final {
+    auto pos_opt = allocation_info_map_.getScopeMap().getMaybeExprPos(expr);
+    if (!pos_opt.has_value()) {
+      // Aliased kir::Allocate exprs are replaced, so the replacements will not
+      // exist in the original ScopeMap. We handle those manually here.
+      position_++;
+    } else {
+      position_ = pos_opt.value();
     }
-    auto n_minus_one = IrBuilder::create<Val>(alignment - 1, DataType::Index);
-    return SimplifyingIrBuilder::bitwiseAndExpr(
-        SimplifyingIrBuilder::addExpr(addr, n_minus_one),
-        SimplifyingIrBuilder::bitwiseNotExpr(n_minus_one));
-  }
 
-  void handle(kir::Allocate* alloc) final {
-    if (alloc->memoryType() != MemoryType::Shared || alloc->alias()) {
-      return;
+    // Check whether this is a first write position for any allocations
+    auto it = first_write_positions_.find(position_);
+    if (it != first_write_positions_.end()) {
+      for (auto alloc_info : it->second) {
+        if (alloc_info->mem_type != MemoryType::Shared) {
+          return;
+        }
+        waiting_to_push_.push_back(alloc_info);
+      }
     }
 
-    const auto buffer_dtype = alloc->buffer()->dtype();
-    const auto dtype_size = dataTypeSize(buffer_dtype);
+    kir::IrVisitor::dispatch(expr);
+  }
+
+  // Reclaim memory whenever we pass an Expr that is known to synchronize the
+  // block
+  void handle(kir::BlockSync* expr) final {
+    reclaimMemory();
+  }
+  void handle(kir::GridSync* expr) final {
+    reclaimMemory();
+  }
+  void handle(ReductionOp* rop) final {
+    const auto output = rop->out()->as<TensorView>();
+    const auto domain = output->domain();
+    if (domain->hasBlockReduction() || domain->hasGridReduction()) {
+      reclaimMemory();
+    }
+  }
+  void handle(WelfordOp* wop) final {
+    const auto output = wop->outAvg()->as<TensorView>();
+    const auto domain = output->domain();
+    if (domain->hasBlockReduction() || domain->hasGridReduction()) {
+      reclaimMemory();
+    }
+  }
+  void handle(kir::CpAsyncWait* expr) final {
```

Comment:
I don't think `CpAsyncWait` implies a block sync.

Reply:
You're right. The corresponding `cpAsyncCg` might have come from before the last popped allocation's "last read", so it's not safe to reclaim memory yet without more information.